spring-kafka-test/src/main/java/org/springframework/kafka/test/EmbeddedKafkaBroker.java

```diff
@@ -732,27 +765,30 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
 
 			@Override
 			public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
-				assigned.set(true);
+				assigned.set(partitions);
 				logger.debug(() -> "partitions assigned: " + partitions);
 			}
 
 		});
 		ConsumerRecords<?, ?> records = null;
 		int n = 0;
-		while (!assigned.get() && n++ < 600) { // NOSONAR magic #
+		while (assigned.get() == null && n++ < 600) { // NOSONAR magic #
 			records = consumer.poll(Duration.ofMillis(100)); // force assignment NOSONAR magic #
 		}
-		if (records != null && records.count() > 0) {
+		if (assigned.get() != null) {
 			final ConsumerRecords<?, ?> theRecords = records;
-			logger.debug(() -> "Records received on initial poll for assignment; re-seeking to beginning; "
-					+ theRecords.partitions().stream()
-					.flatMap(p -> theRecords.records(p).stream())
-					// map to same format as send metadata toString()
-					.map(r -> r.topic() + "-" + r.partition() + "@" + r.offset())
-					.collect(Collectors.toList()));
-			consumer.seekToBeginning(records.partitions());
-		}
-		if (!assigned.get()) {
+			logger.debug(() -> "Partitions assigned "
+					+ assigned.get()
+					+ "; re-seeking to "
+					+ (seekToEnd ? "end; " : "beginning"));
+			if (seekToEnd) {
+				consumer.seekToEnd(assigned.get());
```

Comment:
> 		logger.debug(() -> "Records received on initial poll for assignment; re-seeking to beginning; "
					+ theRecords.partitions().stream()

Is it really correct to `seekToEnd` when we already got this records on a first poll?
Or that doesn't matter and we are not interested in this records any more?

Reply:
>Or that doesn't matter and we are not interested in this records any more?

Correct; the seek was added because, sometimes, the poll that causes the assignment could fetch some records (generally not, but sometimes) and this caused some tests to fail occasionally because we never saw those records. Equally, I suspect, we might not get all records (from all partitions) on this initial poll so, to satisfy the reported problem, we just seek all the assignments (to beginning or end).

This way, the seeks are unconditional, regardless of whether we actually get any records here.

Reply:
Oh! Right. I see now that you have removed that `if (records != null && records.count() > 0) {`.
It is still there in `main` code, so I was confused.

LGTM now. 
Merging...